### PR TITLE
Add `otelcol.processor.groupbyattrs` component

### LINKED ISSUE #224: Add `otelcol.processor.groupbyattrs` component
### Request

Add the OTEL module [groupbyattrsprocessor](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/processor/groupbyattrsprocessor#section-readme).

### Use case

I have an agent scraping prometheus metrics from a kubernetes kubelet, whose metrics I also pass through an `otelcol.processor.k8sattributes` step (and a `otelcol.processor.transform` to turn otel resource attributes to datapoint attributes, [as mentioned in the `otelcol.processor.k8sattributes` docs](https://grafana.com/docs/agent/latest/flow/reference/components/otelcol.processor.k8sattributes/#adding-kubernetes-metadata-to-prometheus-metrics)). What I want is basically what's stated in that doc section, to enrich the kubelet prometheus metrics (with labels like `pod="myapp-59fb8bb44d-6qfgf",namespace="default"`) with metadata that becomes extra labels (like `k8s_deployment_name="myapp"`). The pipeline looks like this:

```text
prometheus.scrape 
  -> otelcol.receiver.prometheus 
    -> otelcol.processor.k8sattributes 
      -> otelcol.processor.transform 
        -> otelcol.exporter.prometheus
```

However, with that pipeline, I get attributes from the k8sattributes processor that _don't match up_ with the labels reported by kubelet. For example, I'll get a datapoint with `namespace="default",k8s_namespace_name="kube-system"` (where `namespace` is from kubelet and `k8s_namespace_name` is from k8sattributes). Also, most datapoints seem "sticky" to particular label values, i.e. many pods across many namespaces will all get a `k8s_namespace_name="kube-system"` label.

What I believe is happening is that in the otel components, all the datapoints are associated with a single Resource for the single `prometheus.scrape` job. So, the k8sattributes step will talk to the Kubernetes API and update the otel Resource, but because the resource is shared, there are conflicts and eventually the last (or the first) write wins and all the other datapoints get those labels.

If that's what's happening, adding the `groupbyattrsprocessor` module should at least let me fix this pipeline so that the k8sattributes enrichment works as expected.

### PR DESCRIPTION
**Pull request by kehindesalaam**
#### PR Description
Adds the `otelcol.processor.groupbyattrs` component which allows grouping of metrics matching specific attributes.

#### Which issue(s) this PR fixes

Fixes #224

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [x] Config converters updated

### CODE REVIEW

**Comment by clayton-cornell** [comment]
@grafana/grafana-alloy-maintainers  Can we do a code review before I review the doc?

File: go.mod
Comment (wildum) [inline go.mod:710]: the version should match the current version of our otel dependency. We update all components at the same time before every release. The current version is `v0.102.1`

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (wildum) [inline internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:20]: the upstream component is in stability beta which is quite stable in opentelemetry terms. I think we can put it directly to PublicPreview or even to GenerallyAvailable. What do you think @ptodev ?

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (wildum) [inline internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:25]: this is the wrong factory

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (wildum) [inline internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:40]: this block should be documented (like the other otel components supporting this block)

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (wildum) [inline internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:48]: I don't think that we need to specify this because it will be empty by default

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (wildum) [inline internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:68]: I don't think that this function is needed, you could return the config directly in Convert()

**Review by wildum — COMMENTED** [review]
Thanks for your contribution! I left a few comments. Let me know when this is ready for a second review :)

**Comment by wildum** [comment]
You will also need to add a config converter for the component. You can follow the pattern used for the other otel components in  https://github.com/grafana/alloy/tree/main/internal/converter/internal/otelcolconvert

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:12]: I'll do that thank you.

File: go.mod
Comment (kehindesalaam) [reply go.mod:710]: I'll make the change

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (kehindesalaam) [reply internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:48]: does it matter if it is nil or non-nil?

Comment (wildum) [reply internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:48]: I checked the Otel code and I don't think that it matters. I see indeed that they initialize it to an empty slice but IMO because it's a config it makes more sense to leave it as nil (in general I would initialize slices as non nil if there is a chance that a value might be added later but this is not the case here).
That being said, it does not really matter in this case. If you think that it feels safer to have it non-nil to match with the default in the otel library I'm fine with it

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:1]:
```suggestion
---
canonical: https://grafana.com/docs/alloy/latest/reference/components/otelcol/otelcol.processor.groupbyattrs/
description: Learn about otelcol.processor.groupbyattrs
title: otelcol.processor.groupbyattrs
---

# otelcol.processor.groupbyattrs
```

We need at least this basic metadata

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:4]:
```suggestion
components and reassociates spans, log records, and metric datapoints to a resource that matches the specified attributes. It groups telemetry data by specified attributes.
```

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:8]:
```suggestion
Collector `groupbyattrs` processor. If necessary, bug reports or feature requests
```

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:12]: There are probably some other changes that will be required here once the other update is applied

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:44]:
```suggestion
`input` | `otelcol.Consumer` |  Accepts `otelcol.Consumer` data for metrics, logs, or traces.
```

I'm not sure about this description. I don't think components can send data to a value... what if we pulled up the extended description. If this works, we could remove lines 46 and 47.

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:88]:
```suggestion
- The data points with different data types aren't merged under the same metric. For example, a gauge and sum metric would not be merged.
- The data points without the specified keys remain under their respective resources.
- New resources inherit the attributes of the original resource and the specified attributes in the keys array.
- The grouping attributes in the keys array are removed from the output metrics.
- If the keys array is empty, the processor performs compaction and reassociates all spans with matching Resource and InstrumentationLibrary.
```

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:83]: Non-blocking commen.
I'm not sure that this info is best grouped under a `Notes` heading.  I'll leave it as-is... just noting it here in the review that this is not used elsewhere, and it might be better to rethink how this information is presented. It's something we can put off to a future PR.

**Review by clayton-cornell — COMMENTED** [review]
Some doc suggestions

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:44]: I'll look into that

Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md]: There is no extended description for this section. All the files have copies of the same thing. It is a good ideaa to make this part into a separate file. I would assume it is the same for all otelcol processors

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:18]: done

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (kehindesalaam) [reply internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go]: I made it GenerallyAvailable

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (kehindesalaam) [reply internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:39]: I added a doc for this

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (kehindesalaam) [reply internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go]: removed

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:89]: I moved the last note as part of the documentation for the Keys attribute

**Comment by kehindesalaam** [comment]
@wildum could you please take another look?

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:41]:
```suggestion
| `keys`          | `array(string)`   | Keys that will be used to group the spans, log records, or metric data points together. |         | no       |
```

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:18]:
```suggestion
We recommend you use the groupbyattrs processor together with [otelcol.processor.batch][], as a consecutive step. This will reduce the fragmentation of data by grouping records together under the matching Resource/Instrumentation Library.
```

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:78]:
```suggestion
`otelcol.processor.groupbyattrs` doesn't expose any component-specific debug
```

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:83]:
```suggestion
`otelcol.processor.groupbyattrs` doesn't expose any component-specific debug metrics.
```

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (wildum) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:10]: I would maybe remove this sentence "It groups telemetry data by specified attributes.". It does not add a lot of info. If we think that this is not clear enough, maybe an example would help

**Review by wildum — COMMENTED** [review]
the code looks good to me, I just have some little feedback on the doc. I found that the examples in the original doc were helpful to understand how the processor work. Should we add some examples or is it already clear enough? @ptodev I think you also wanted to give some feedback on this PR?

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (ptodev) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:109]:
```suggestion
### Grouping metrics

Consider the below metrics, all originally associated to the same Resource:

``
Resource {host.name="localhost",source="prom"}
  Metric "gauge-1" (GAUGE)
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-B",id="eth0"}
  Metric "gauge-1" (GAUGE) // Identical to previous Metric
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-B",id="eth0"}
  Metric "mixed-type" (GAUGE)
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-B",id="eth0"}
  Metric "mixed-type" (SUM)
    DataPoint {host.name="host-A",id="eth0"}
    DataPoint {host.name="host-A",id="eth0"}
  Metric "dont-move" (Gauge)
    DataPoint {id="eth0"}
``

With the below configuration, the groupbyattrs will re-associate the metrics with either `host-A` or `host-B`, based on the value of the `host.name` attribute.

``alloy
otelcol.processor.groupbyattrs "default" {
  keys = [ "host.name" ]

  output {
    metrics = [otelcol.exporter.otlp.default.input]
  }
}
``

The output of the processor will therefore be:

``
Resource {host.name="localhost",source="prom"}
  Metric "dont-move" (Gauge)
    DataPoint {id="eth0"}

Resource {host.name="host-A",source="prom"}
  Metric "gauge-1"
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
  Metric "mixed-type" (GAUGE)
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
  Metric "mixed-type" (SUM)
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}

Resource {host.name="host-B",source="prom"}
  Metric "gauge-1"
    DataPoint {id="eth0"}
    DataPoint {id="eth0"}
  Metric "mixed-type" (GAUGE)
    DataPoint {id="eth0"}
``

This output demonstrates how `otelcol.processor.groupbyattrs` works in various situations:

- The DataPoints for the `gauge-1` (GAUGE) metric were originally split under 2 Metric instances and have been merged in the output.
- The DataPoints of the `mixed-type` (GAUGE) and `mixed-type` (SUM) metrics have not been merged under the same Metric, because their DataType is different.
- The `dont-move` metric DataPoints don't have a `host.name` attribute and therefore remained under the original Resource.
- The new Resources inherited the attributes from the original Resource (`source="prom"`), plus the specified attributes from the processed metrics (`host.name="host-A"` or `host.name="host-B"`).
- The specified "grouping" attributes that are set on the new Resources are also removed from the metric DataPoints.
- While not shown in the above example, the processor also merges collections of records under matching InstrumentationLibrary.
```

@clayton-cornell does this look ok? I copied it from the [upstream docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.105.0/processor/groupbyattrsprocessor/README.md#grouping-metrics).

@kehindesalaam Note that you'd need to replace the `` with ```. I can't do it in the GitHub suggestion, because I can't escape the tick symbol inside suggestions.

File: internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go
Comment (ptodev) [inline internal/component/otelcol/processor/groupbyattrs/groupbyattrs.go:53]:
```suggestion
	*args = Arguments{
		Keys: []string{},
	}
```

I think this could avoid issues with shallow copying of slices.

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:109]: @ptodev You can use codefence markdown in suggestions.. just use 4 backticks to wrap the suggestion block and the standard 3 backticks for code bloack.

Instead of `the below`, we can use `the following` (better grammatically).  I'll look at the info-flow a little deeper once it's applied and do another pass on the docs.

We have to be careful with 1:1 copied text from upstream - we should either directly link to it or note it's source (per whatever license upstream is using) if we copy it.

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:10]: I would remove that

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:109]: I'll make the modification and also the adjustment Clayton suggested.

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:1]: I'll run this as well.

**Comment by kehindesalaam** [comment]
Thanks for the review team. I'm sorry the second set of changes came a bit later. I have been busy.

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (ptodev) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:41]:
```suggestion
| `keys`          | `list(string)`    | Keys that will be used to group the spans, log records or metric data points together. |         | no       |
```

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (ptodev) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:156]:
````suggestion

### Compaction

Sometimes telemetry data can become fragmented due to multiple duplicated ResourceSpans/ResourceLogs/ResourceMetrics objects.
This leads to additional memory consumption, increased processing costs, inefficient serialization and increase of the export requests. 
In such situations, `otelcol.processor.groupbyattrs` can be used to compact the data with matching Resource and InstrumentationLibrary properties.

For example, consider this input data:

```
Resource {host.name="localhost"}
  InstrumentationLibrary {name="MyLibrary"}
  Spans
    Span {span_id=1, ...}
  InstrumentationLibrary {name="OtherLibrary"}
  Spans
    Span {span_id=2, ...}
    
Resource {host.name="localhost"}
  InstrumentationLibrary {name="MyLibrary"}
  Spans
    Span {span_id=3, ...}
    
Resource {host.name="localhost"}
  InstrumentationLibrary {name="MyLibrary"}
  Spans
    Span {span_id=4, ...}
    
Resource {host.name="otherhost"}
  InstrumentationLibrary {name="MyLibrary"}
  Spans
    Span {span_id=5, ...}
```

You can use `otelcol.processor.groupbyattrs` with its default configuration to compact the data: 
```alloy
otelcol.processor.groupbyattrs "default" {
  output {
    metrics = [otelcol.exporter.otlp.default.input]
  }
}
```

The output will be:

```
Resource {host.name="localhost"}
  InstrumentationLibrary {name="MyLibrary"}
  Spans
    Span {span_id=1, ...}
    Span {span_id=3, ...}
    Span {span_id=4, ...}
  InstrumentationLibrary {name="OtherLibrary"}
  Spans
    Span {span_id=2, ...}

Resource {host.name="otherhost"}
  InstrumentationLibrary {name="MyLibrary"}
  Spans
    Span {span_id=5, ...}
```
````

This example was taken from the [OTel docs](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/processor/groupbyattrsprocessor#section-readme).

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (ptodev) [inline docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:157]:
```suggestion
[otelcol.processor.batch]: ../otelcol.processor.batch/
```

File: internal/converter/internal/otelcolconvert/testdata/groupbyattrs.alloy
Comment (ptodev) [inline internal/converter/internal/otelcolconvert/testdata/groupbyattrs.alloy:1]: @kehindesalaam The test related to this file is failing. I think you'll need to rebase your branch with `main` in order to fix it.

**Review by ptodev — COMMENTED** [review]
Thank you, @kehindesalaam! When @clayton-cornell approves it and when the CI is green, I'm happy for the PR to be merged!

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (clayton-cornell) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:156]: @ptodev We have to be super careful about taking (copy/pasting) content - any content - from external sources including the OTel.  We are probably OK here, but this is something we need to keep in mind when pulling info from external sources :-)  https://developers.google.com/style/other-sources

Comment (clayton-cornell) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:156]: Just t be clear, this suggestion should be OK as-is.

File: internal/converter/internal/otelcolconvert/testdata/groupbyattrs.alloy
Comment (kehindesalaam) [reply internal/converter/internal/otelcolconvert/testdata/groupbyattrs.alloy:1]: rebased

File: docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md
Comment (ptodev) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:156]: Hi, @kehindesalaam when the section from this suggestion is added, I'll give the PR a final look and will likely ✅  it. Thank you for your contributions!

Comment (kehindesalaam) [reply docs/sources/reference/components/otelcol/otelcol.processor.groupbyattrs.md:156]: I missed this 😅

**Review by Allzz77123 — APPROVED** [review]

**Review by ptodev — APPROVED** [review]
Thank you! Looks good to me. 🚢  I just pushed a commit on your branch with a few minor docs fixes. By the way, you can visualise the docs in a browser by running these commands:
```
cd docs
make docs
```
I find this very useful when I want to make sure everything gets rendered appropriately.

**Review by clayton-cornell — APPROVED** [review]
Doc are ok

**Pull request merged**
